program/src/processor.rs

```diff
@@ -351,6 +355,104 @@ pub fn process_stake_deposit(
+        &accounts.validator_vote_account.key.to_bytes(),
+        VALIDATOR_UNSTAKE_ACCOUNT,
+        &validator.entry.unstake_seeds.end.to_le_bytes()[..],
+        &[destination_bump_seed],
+    ];
+
+    split_stake_account(
+        accounts.lido.key,
+        &lido,
+        &SplitStakeAccounts {
+            source_stake_account: accounts.source_stake_account,
+            destination_stake_account: accounts.destination_stake_account,
+            authority: accounts.stake_authority,
+            system_program: accounts.system_program,
+            stake_program: accounts.stake_program,
+        },
+        amount,
+        &[&seeds],
+    )?;
+
+    let deactivate_stake_instruction = solana_program::stake::instruction::deactivate_stake(
+        accounts.destination_stake_account.key,
+        accounts.stake_authority.key,
+    );
+
+    // Deactivates the stake. After the stake has become inactive, the Lamports
+    // on this stake account need to go back to the reserve account by using
+    // another instruction.
+    invoke_signed(
+        &deactivate_stake_instruction,
+        &[
+            accounts.destination_stake_account.clone(),
+            accounts.sysvar_clock.clone(),
+            accounts.stake_authority.clone(),
+            accounts.stake_program.clone(),
+        ],
+        &[&[
+            &accounts.lido.key.to_bytes(),
+            STAKE_AUTHORITY,
+            &[lido.stake_authority_bump_seed],
+        ]],
+    )?;
+
+    let validator = lido
+        .validators
+        .get_mut(accounts.validator_vote_account.key)?;
+    // Increase the `unstake_accounts_balance` by `amount`.
+    validator.entry.unstake_accounts_balance = (validator.entry.unstake_accounts_balance + amount)?;
+    msg!(
+        "VALIDATOR ACTIVE: {}, STAKE ACCOUNT:{}",
+        validator.entry.active,
+        validator.entry.stake_accounts_balance
+    );
+    if validator.entry.active {
+        if (validator.entry.stake_accounts_balance - amount)? < MINIMUM_STAKE_ACCOUNT_BALANCE {
+            msg!(
+            "Unstake operation will leave the active validator with {}, less than the minimum balance {}.\\
+            Only inactive validators can fall below the limit.",
+            validator.entry.stake_accounts_balance,
+            MINIMUM_STAKE_ACCOUNT_BALANCE
+        );
+            return Err(LidoError::InvalidAmount.into());
+        }
+    } else {
+        let full_amount = (validator.entry.stake_accounts_balance
+            - validator.entry.unstake_accounts_balance)
+            .expect("Unstake accounts balance is not greater than Stake accounts balance.");
+        if full_amount != amount {
+            msg!(
+                "An inactive validator must have all its stake withdrawn. Tried\\
+                to withdraw {}, Should withdraw {} instead.",
+                amount,
+                full_amount,
+            );
+            return Err(LidoError::InvalidAmount.into());
+        }
```

Comment:
In this case, we also need to bump `stake_accounts.begin`. Otherwise the range `begin..end` will contain non-existing accounts. And maybe we can check then that `stake_accounts.begin == stake_accounts.end`, i.e. we removed the last stake account that the validator had.